File: src/main/kotlin/customfit/ai/kotlinclient/network/ConfigFetcher.kt
package customfit.ai.kotlinclient.network

import customfit.ai.kotlinclient.config.change.CFConfigChangeManager
import customfit.ai.kotlinclient.constants.CFConstants
import customfit.ai.kotlinclient.config.core.CFConfig
import customfit.ai.kotlinclient.core.error.CFResult
import customfit.ai.kotlinclient.core.error.ErrorHandler
import customfit.ai.kotlinclient.core.model.CFUser
import customfit.ai.kotlinclient.core.util.RetryUtil.withRetry
import customfit.ai.kotlinclient.logging.Timber
import java.util.concurrent.atomic.AtomicBoolean
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.SerializationException
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*

/** Handles fetching configuration from the CustomFit API with support for offline mode */
class ConfigFetcher(
        private val httpClient: HttpClient,
        private val cfConfig: CFConfig,
        private val user: CFUser
) {
    companion object {
        private const val SOURCE = "ConfigFetcher"
    }
    
    private val offlineMode = AtomicBoolean(false)
    private val fetchMutex = Mutex()
    private var lastConfigMap: Map<String, Any>? = null
    private val mutex = Mutex()
    private var lastFetchTime: Long = 0

    /** Returns whether the client is in offline mode */
    fun isOffline(): Boolean = offlineMode.get()

    /**
     * Sets the offline mode status
     *
     * @param offline true to enable offline mode, false to disable
     */
    fun setOffline(offline: Boolean) {
        offlineMode.set(offline)
        Timber.i("ConfigFetcher offline mode set to: $offline")
    }

    /**
     * Fetches configuration from the API with improved error handling
     *
     * @param lastModified Optional last-modified header value for conditional requests
     * @param etag Optional ETag header value for conditional requests
     * @return CFResult containing configuration map or error details
     */
    suspend fun fetchConfig(lastModified: String? = null, etag: String? = null): CFResult<Map<String, Any>> {
        // Don't fetch if in offline mode
        if (isOffline()) {
            Timber.d("Not fetching config because client is in offline mode")
            return CFResult.error(
                "Client is in offline mode",
                category = ErrorHandler.ErrorCategory.NETWORK
            )
        }

        return fetchMutex.withLock {
            try {
                val url = "${CFConstants.Api.BASE_API_URL}${CFConstants.Api.USER_CONFIGS_PATH}?cfenc=${cfConfig.clientKey}"
                
                // Build payload using kotlinx.serialization
                val jsonObject = buildJsonObject {
                    put(
                        "user",
                        buildJsonObject {
                            user.toUserMap().forEach { (k, v) -> put(k, anyToJsonElement(v)) }
                        }
                    )
                    put("include_only_features_flags", JsonPrimitive(true))
                }
                val payload = Json.encodeToString(jsonObject)

                Timber.i("API POLL: Fetching config from URL: $url")
                Timber.d("Config fetch payload: $payload")

                val headers = mutableMapOf<String, String>(
                    CFConstants.Http.HEADER_CONTENT_TYPE to CFConstants.Http.CONTENT_TYPE_JSON
                )
                
                // Add conditional request headers if available
                lastModified?.let {
                    headers[CFConstants.Http.HEADER_IF_MODIFIED_SINCE] = it
                    Timber.i("API POLL: Using If-Modified-Since: $it")
                }
                
                etag?.let {
                    headers[CFConstants.Http.HEADER_IF_NONE_MATCH] = it
                    Timber.i("API POLL: Using If-None-Match: $it")
                }

                // Use the updated HttpClient that returns CFResult
                val responseBodyResult = suspendGetResponseBody(url, headers, payload)
                
                if (responseBodyResult.isFailure) {
                    val exception = responseBodyResult.exceptionOrNull()
                    ErrorHandler.handleError(
                        "Failed to fetch configuration body", 
                        SOURCE,
                        ErrorHandler.ErrorCategory.NETWORK,
                        ErrorHandler.ErrorSeverity.HIGH
                    )
                    return@withLock CFResult.error(
                        "Failed to fetch configuration body: ${exception?.message ?: "Unknown error"}",
                        exception,
                        category = ErrorHandler.ErrorCategory.NETWORK
                    )
                }
                
                val responseBody = responseBodyResult.getOrNull()
                if (responseBody == null) {
                    ErrorHandler.handleError(
                        "Failed to fetch configuration body (empty response)", 
                        SOURCE,
                        ErrorHandler.ErrorCategory.NETWORK,
                        ErrorHandler.ErrorSeverity.HIGH
                    )
                    return@withLock CFResult.error(
                        "Failed to fetch configuration body (empty response)",
                        category = ErrorHandler.ErrorCategory.NETWORK
                    )
                }
                
                Timber.i("API POLL: Successfully fetched config, response size: ${responseBody.length} bytes")
                
                // Process configuration response
                return@withLock processConfigResponse(responseBody)
            } catch (e: Exception) {
                val category = when (e) {
                    is SerializationException -> ErrorHandler.ErrorCategory.SERIALIZATION
                    else -> ErrorHandler.ErrorCategory.INTERNAL
                }
                
                Timber.e("API POLL: Error fetching configuration: ${e.message}")
                
                ErrorHandler.handleException(
                    e,
                    "Error fetching configuration",
                    SOURCE,
                    ErrorHandler.ErrorSeverity.HIGH
                )
                
                CFResult.error(
                    "Error fetching configuration: ${e.message}",
                    e,
                    category = category
                )
            }
        }
    }
    
    /**
     * Helper method to get the response body from a URL
     */
    private suspend fun suspendGetResponseBody(url: String, headers: Map<String, String>, payload: String): Result<String?> {
        return try {
            val result = httpClient.performRequest(url, "POST", headers, payload) { conn ->
                val responseCode = conn.responseCode
                if (responseCode == java.net.HttpURLConnection.HTTP_OK) {
                    conn.inputStream.bufferedReader().readText()
                } else {
                    Timber.w("Failed to fetch config from $url: $responseCode")
                    null
                }
            }
            Result.success(result)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * Process the configuration response, with improved error handling
     *
     * @param jsonResponse The JSON response string from the API
     * @return CFResult containing the processed config map or error details
     */
    private fun processConfigResponse(jsonResponse: String): CFResult<Map<String, Any>> {
        val finalConfigMap = mutableMapOf<String, Any>()

        try {
            // Parse the entire response string into a JsonElement
            val responseElement = Json.parseToJsonElement(jsonResponse)
            if (responseElement !is JsonObject) {
                val message = "Response is not a JSON object"
                ErrorHandler.handleError(
                    message,
                    SOURCE,
                    ErrorHandler.ErrorCategory.SERIALIZATION,
                    ErrorHandler.ErrorSeverity.HIGH
                )
                return CFResult.error(
                    message,
                    category = ErrorHandler.ErrorCategory.SERIALIZATION
                )
            }
            
            val responseJson = responseElement.jsonObject
            val configsJson = responseJson["configs"]?.jsonObject

            if (configsJson == null) {
                val message = "No 'configs' object found in the response"
                ErrorHandler.handleError(
                    message,
                    SOURCE,
                    ErrorHandler.ErrorCategory.VALIDATION,
                    ErrorHandler.ErrorSeverity.MEDIUM
                )
                return CFResult.success(emptyMap())
            }

            // Iterate through each config entry
            configsJson.entries.forEach { (key, configElement) ->
                try {
                    if (configElement !is JsonObject) {
                        ErrorHandler.handleError(
                            "Config entry for '$key' is not a JSON object",
                            SOURCE,
                            ErrorHandler.ErrorCategory.SERIALIZATION,
                            ErrorHandler.ErrorSeverity.MEDIUM
                        )
                        return@forEach
                    }
                    
                    val configObject = configElement.jsonObject
                    val experienceObject = configObject["experience_behaviour_response"]?.jsonObject

                    // Convert the config JsonObject to a mutable map
                    val flattenedMap = jsonObjectToMap(configObject).toMutableMap()

                    // Remove the nested object itself (it will be merged)
                    flattenedMap.remove("experience_behaviour_response")

                    // Merge fields from the nested experience object if it exists
                    experienceObject?.let { flattenedMap.putAll(jsonObjectToMap(it)) }

                    // Store the flattened map, ensuring Any? values are handled or filtered if needed
                    @Suppress("UNCHECKED_CAST")
                    finalConfigMap[key] =
                            flattenedMap.filterValues { it != null } as Map<String, Any>
                } catch (e: Exception) {
                    ErrorHandler.handleException(
                        e,
                        "Error processing individual config key '$key'",
                        SOURCE,
                        ErrorHandler.ErrorSeverity.MEDIUM
                    )
                }
            }

            // Notify observers of config changes
            if (finalConfigMap != lastConfigMap) {
                CFConfigChangeManager.notifyObservers(finalConfigMap, lastConfigMap)
                lastConfigMap = finalConfigMap
                lastFetchTime = System.currentTimeMillis()
            }
            
            // Log config details with both keys and values
            Timber.d("Config keys: ${finalConfigMap.keys}")
            
            // Print each config key and its variation value only
            finalConfigMap.forEach { (key, value) ->
                when (value) {
                    is Map<*, *> -> {
                        val variation = value["variation"]
                        Timber.d("$key: $variation")
                    }
                    else -> {
                        Timber.d("$key: $value")
                    }
                }
            }
            
            // Keep existing hero_text debug logging for backward compatibility
            val heroText = (finalConfigMap["hero_text"] as? Map<String, Any>)?.get("variation") as? String
            if (heroText != null) {
                Timber.d("Hero text if present: $heroText")
            }
            
            return CFResult.success(finalConfigMap)
        } catch (e: Exception) {
            ErrorHandler.handleException(
                e,
                "Error parsing configuration response",
                SOURCE,
                ErrorHandler.ErrorSeverity.HIGH
            )
            
            return CFResult.error(
                "Error parsing configuration response: ${e.message}",
                e,
                category = ErrorHandler.ErrorCategory.SERIALIZATION
            )
        }
    }

    // --- kotlinx.serialization based helpers ---

    /** Converts a JsonObject to a Map<String, Any?>. */
    private fun jsonObjectToMap(jsonObject: JsonObject): Map<String, Any?> {
        return jsonObject.mapValues { jsonElementToValue(it.value) }
    }

    /** Converts a JsonArray to a List<Any?>. */
    private fun jsonArrayToList(jsonArray: JsonArray): List<Any?> {
        return jsonArray.map { jsonElementToValue(it) }
    }

    /** Recursively converts a JsonElement to a Kotlin primitive, Map, or List. */
    private fun jsonElementToValue(element: JsonElement?): Any? {
        return when (element) {
            is JsonNull -> null
            is JsonPrimitive ->
                    when {
                        element.isString -> element.content
                        element.booleanOrNull != null -> element.boolean
                        element.longOrNull != null -> element.long // Prioritize Long
                        element.doubleOrNull != null -> element.double // Then Double
                        else -> element.content // Fallback
                    }
            is JsonObject -> jsonObjectToMap(element)
            is JsonArray -> jsonArrayToList(element)
            null -> null
        }
    }

    /**
     * Fetches metadata from a URL with improved error handling
     * Optimized to use HEAD requests first to minimize bandwidth usage
     *
     * @param url The URL to fetch metadata from
     * @return CFResult containing metadata headers or error details
     */
    suspend fun fetchMetadata(url: String): CFResult<Map<String, String>> {
        if (isOffline()) {
            Timber.d("Not fetching metadata because client is in offline mode")
            return CFResult.error(
                "Client is in offline mode",
                category = ErrorHandler.ErrorCategory.NETWORK
            )
        }

        try {
            // First try a lightweight HEAD request
            Timber.i("API POLL: Fetch metadata strategy - First trying HEAD request: $url")
            val headResult = httpClient.makeHeadRequest(url)
            
            if (headResult is CFResult.Success) {
                Timber.i("API POLL: HEAD request successful, using result: ${headResult.data}")
                return headResult
            } else {
                // If HEAD fails, fall back to the original GET method
                Timber.i("API POLL: HEAD request failed, falling back to GET: $url")
                val getResult = httpClient.fetchMetadata(url)
                if (getResult is CFResult.Success) {
                    Timber.i("API POLL: Fallback GET successful: ${getResult.data}")
                } else {
                    Timber.w("API POLL: Both HEAD and GET failed for $url")
                }
                return getResult
            }
        } catch (e: Exception) {
            Timber.e("API POLL: Exception during metadata fetch attempts: ${e.message}")
            ErrorHandler.handleException(
                e,
                "Error fetching metadata from $url",
                SOURCE,
                ErrorHandler.ErrorSeverity.HIGH
            )
            
            return CFResult.error(
                "Error fetching metadata: ${e.message}",
                e,
                category = ErrorHandler.ErrorCategory.NETWORK
            )
        }
    }

    /**
     * Fetches complete SDK settings from a URL, including both metadata headers and the full settings object
     * This is preferred over fetchMetadata when you need to process the actual settings content
     *
     * @param url The URL to fetch SDK settings from
     * @return CFResult containing both headers and parsed SdkSettings object, or error details
     */
    suspend fun fetchSdkSettingsWithMetadata(url: String): CFResult<Pair<Map<String, String>, customfit.ai.kotlinclient.core.model.SdkSettings?>> {
        if (isOffline()) {
            Timber.d("Not fetching SDK settings because client is in offline mode")
            return CFResult.error(
                "Client is in offline mode",
                category = ErrorHandler.ErrorCategory.NETWORK
            )
        }

        try {
            // Always use GET for this method since we need the full response body
            Timber.i("API POLL: Fetching full SDK settings with GET: $url")
            
            val result = httpClient.performRequest(url, "GET", emptyMap(), null) { conn ->
                val responseCode = conn.responseCode
                if (responseCode == java.net.HttpURLConnection.HTTP_OK) {
                    // Extract headers
                    val headers = mutableMapOf<String, String>()
                    headers[CFConstants.Http.HEADER_LAST_MODIFIED] = conn.getHeaderField("Last-Modified") ?: ""
                    headers[CFConstants.Http.HEADER_ETAG] = conn.getHeaderField("ETag") ?: ""
                    
                    // Read the response body
                    val body = conn.inputStream.bufferedReader().readText()
                    Timber.i("API POLL: SDK settings response received, size: ${body.length} bytes")
                    
                    // Return both headers and body
                    Pair(headers, body)
                } else {
                    Timber.w("API POLL: Failed to fetch SDK settings from $url: $responseCode")
                    null
                }
            }
            
            if (result == null) {
                return CFResult.error(
                    "Failed to fetch SDK settings",
                    category = ErrorHandler.ErrorCategory.NETWORK
                )
            }
            
            val (headers, body) = result
            
            // Parse the body into SdkSettings
            val sdkSettings = parseSdkSettings(body)
            
            if (sdkSettings != null) {
                Timber.i("API POLL: SDK settings parsed successfully, account enabled: ${sdkSettings.cf_account_enabled}")
            } else {
                Timber.w("API POLL: Failed to parse SDK settings response")
            }
            
            return CFResult.success(Pair(headers, sdkSettings))
        } catch (e: Exception) {
            Timber.e("API POLL: Exception during SDK settings fetch: ${e.message}")
            ErrorHandler.handleException(
                e,
                "Error fetching SDK settings from $url",
                SOURCE,
                ErrorHandler.ErrorSeverity.HIGH
            )
            
            return CFResult.error(
                "Error fetching SDK settings: ${e.message}",
                e,
                category = ErrorHandler.ErrorCategory.NETWORK
            )
        }
    }
    
    /**
     * Parse SDK settings JSON into a simplified SdkSettings object
     * Only extracts the essential fields needed for core functionality
     *
     * @param jsonString The JSON string to parse
     * @return SdkSettings object if successful, null otherwise
     */
    private fun parseSdkSettings(jsonString: String): customfit.ai.kotlinclient.core.model.SdkSettings? {
        return try {
            val jsonObject = Json.parseToJsonElement(jsonString).jsonObject
            
            // Extract only the essential boolean values we need with fallbacks
            val cfAccountEnabled = jsonObject["cf_account_enabled"]?.jsonPrimitive?.booleanOrNull ?: true
            val cfSkipSdk = jsonObject["cf_skip_sdk"]?.jsonPrimitive?.booleanOrNull ?: false
            
            // Log that we're using a simplified version
            Timber.d("Parsing SDK settings with simplified model (only essential fields)")
            
            // Create and return simplified SdkSettings object with just the fields we need
            customfit.ai.kotlinclient.core.model.SdkSettings(
                cf_account_enabled = cfAccountEnabled,
                cf_skip_sdk = cfSkipSdk
            )
        } catch (e: Exception) {
            Timber.e(e, "Failed to parse SDK settings: ${e.message}")
            null
        }
    }

    /** Helper function to convert Any to JsonElement */
    private fun anyToJsonElement(value: Any?): JsonElement = when (value) {
        null -> JsonNull
        is String -> JsonPrimitive(value)
        is Number -> JsonPrimitive(value)
        is Boolean -> JsonPrimitive(value)
        is Map<*, *> -> buildJsonObject {
            value.forEach { (k, v) ->
                if (k is String) {
                    put(k, anyToJsonElement(v))
                }
            }
        }
        is List<*> -> buildJsonArray {
            value.forEach { item ->
                add(anyToJsonElement(item))
            }
        }
        else -> JsonPrimitive(value.toString())
    }
}

File: customfit-flutter-client-sdk/lib/src/network/config_fetcher.dart
import 'dart:async';
import 'package:customfit_flutter_client_sdk/src/core/error/cf_result.dart';
import 'package:dio/dio.dart';
import 'dart:convert';

import '../core/model/cf_user.dart';
import '../core/error/error_category.dart';
import '../core/error/error_severity.dart';
import '../core/error/error_handler.dart';
import '../constants/cf_constants.dart';
import '../../logging/logger.dart';
import 'http_client.dart';
import '../config/core/cf_config.dart';

/// Handles fetching configuration from the CustomFit API with support for offline mode
class ConfigFetcher {
  static const String _source = "ConfigFetcher";

  final HttpClient _httpClient;
  final CFConfig _config;
  final CFUser _user;

  bool _offlineMode = false;
  Completer<void> _fetchMutex = Completer<void>();
  Map<String, dynamic>? _lastConfigMap;
  final _mutex = Completer<void>();
  int _lastFetchTime = 0;

  // Store metadata headers for conditional requests
  String? _lastModified;
  String? _lastEtag;

  ConfigFetcher(this._httpClient, this._config, this._user) {
    _fetchMutex.complete();
    _mutex.complete();
    Logger.d('ConfigFetcher initialized with user: ${_user.userCustomerId}');
  }

  /// Returns whether the client is in offline mode
  bool isOffline() => _offlineMode;

  /// Sets the offline mode status
  void setOffline(bool offline) {
    _offlineMode = offline;
    Logger.i("ConfigFetcher offline mode set to: $offline");
  }

  /// Fetches configuration from the API with improved error handling
  Future<bool> fetchConfig({String? lastModified}) async {
    // Don't fetch if in offline mode
    if (isOffline()) {
      Logger.d("Not fetching config because client is in offline mode");
      return false;
    }

    // Create a new mutex if needed
    final completer = Completer<void>();
    final oldCompleter = _fetchMutex;
    if (oldCompleter.isCompleted) {
      _fetchMutex = completer;
    } else {
      Logger.d("Fetch already in progress, waiting...");
      await _fetchMutex.future.timeout(
        const Duration(seconds: 5),
        onTimeout: () {
          Logger.w("Timed out waiting for previous fetch to complete");
          // Continue anyway, but create a new mutex
          _fetchMutex = completer;
        },
      );
    }

    try {
      final url =
          "${CFConstants.api.baseApiUrl}${CFConstants.api.userConfigsPath}?cfenc=${_config.clientKey}";

      Logger.i("API POLL: Fetching config from URL: $url");
      if (lastModified != null) {
        Logger.i("API POLL: Using If-Modified-Since: $lastModified");
      }

      // Build payload
      final jsonObject = {
        "user": _user.toMap(),
        "include_only_features_flags": true,
      };
      final payload = jsonEncode(jsonObject);

      Logger.d("Config fetch payload size: ${payload.length} bytes");

      final headers = <String, String>{
        CFConstants.http.headerContentType: CFConstants.http.contentTypeJson,
      };

      // Add If-Modified-Since header if available (match Kotlin)
      if (lastModified != null) {
        headers[CFConstants.http.headerIfModifiedSince] = lastModified;
      }

      // Add timeout to the HTTP request
      final result = await _httpClient
          .post(
        url,
        data: payload,
        options: Options(
          headers: headers,
          sendTimeout: const Duration(milliseconds: 10000),
          receiveTimeout: const Duration(milliseconds: 10000),
        ),
      )
          .timeout(
        const Duration(seconds: 10),
        onTimeout: () {
          Logger.w("API POLL: Request timed out after 10 seconds");
          return CFResult.error("Request timed out",
              category: ErrorCategory.network);
        },
      );

      // Handle 304 Not Modified (match Kotlin)
      if (result.getStatusCode() == 304) {
        Logger.i("API POLL: Configs not modified (304), using cached configs");
        completer.complete();
        return true;
      }

      if (!result.isSuccess) {
        Logger.w("API POLL: Failed to fetch: ${result.getErrorMessage()}");
        ErrorHandler.handleError(
          "Failed to fetch configuration: ${result.getErrorMessage()}",
          source: _source,
          category: ErrorCategory.network,
          severity: ErrorSeverity.high,
        );
        completer.complete();
        return false;
      }

      final responseBody = result.getOrNull();
      if (responseBody == null) {
        Logger.w("API POLL: Empty response received");
        ErrorHandler.handleError(
          "Empty configuration response",
          source: _source,
          category: ErrorCategory.network,
          severity: ErrorSeverity.high,
        );
        completer.complete();
        return false;
      }

      Logger.i(
          "API POLL: Successfully fetched config, response size: ${responseBody.toString().length} bytes");
      final handled = _handleConfigResponse(responseBody);
      Logger.i("API POLL: Handled response successfully: $handled");
      completer.complete();
      return handled;
    } catch (e) {
      Logger.e("API POLL: Error fetching configuration: ${e.toString()}");
      ErrorHandler.handleException(
        e,
        "Error fetching configuration",
        source: _source,
        severity: ErrorSeverity.high,
      );
      completer.complete();
      return false;
    }
  }

  /// Handle different response types
  bool _handleConfigResponse(dynamic responseBody) {
    // Direct boolean responses (e.g., 304 Not Modified)
    if (responseBody is bool) {
      return responseBody;
    }

    try {
      // String response (JSON)
      if (responseBody is String) {
        final processResult = _processConfigResponse(responseBody);
        return processResult.isSuccess;
      }

      // Map response
      else if (responseBody is Map) {
        final jsonStr = jsonEncode(responseBody);
        final processResult = _processConfigResponse(jsonStr);
        return processResult.isSuccess;
      }

      // Unexpected response type
      else {
        Logger.w("Unexpected response type: ${responseBody.runtimeType}");
        return false;
      }
    } catch (e) {
      ErrorHandler.handleException(
        e,
        "Error processing response",
        source: _source,
        severity: ErrorSeverity.high,
      );
      return false;
    }
  }

  /// Process the configuration response, with improved error handling
  CFResult<Map<String, dynamic>> _processConfigResponse(String jsonResponse) {
    final finalConfigMap = <String, dynamic>{};

    try {
      // Parse the entire response string into a Map
      final responseMap = jsonDecode(jsonResponse) as Map<String, dynamic>;

      final configsJson = responseMap['configs'] as Map<String, dynamic>?;

      if (configsJson == null) {
        const message = "No 'configs' object found in the response";
        ErrorHandler.handleError(
          message,
          source: _source,
          category: ErrorCategory.validation,
          severity: ErrorSeverity.medium,
        );
        return CFResult.success(<String, dynamic>{});
      }

      // Iterate through each config entry
      configsJson.forEach((key, configElement) {
        try {
          if (configElement is! Map<String, dynamic>) {
            ErrorHandler.handleError(
              "Config entry for '$key' is not a JSON object",
              source: _source,
              category: ErrorCategory.serialization,
              severity: ErrorSeverity.medium,
            );
            return;
          }

          final configObject = configElement;
          final experienceObject = configObject['experience_behaviour_response']
              as Map<String, dynamic>?;

          // Convert the config object to a map
          final flattenedMap = Map<String, dynamic>.from(configObject);

          // Remove the nested object itself (it will be merged)
          flattenedMap.remove('experience_behaviour_response');

          // Merge fields from the nested experience object if it exists
          if (experienceObject != null) {
            flattenedMap.addAll(experienceObject);
          }

          // Store the flattened map, filtering out null values
          finalConfigMap[key] = Map<String, dynamic>.from(
            flattenedMap..removeWhere((_, value) => value == null),
          );
        } catch (e) {
          ErrorHandler.handleException(
            e,
            "Error processing individual config key '$key'",
            source: _source,
            severity: ErrorSeverity.medium,
          );
        }
      });

      // Notify observers of config changes
      if (finalConfigMap != _lastConfigMap) {
        _lastConfigMap = finalConfigMap;
        _lastFetchTime = DateTime.now().millisecondsSinceEpoch;
      }

      // Log config keys
      Logger.d("Config keys: ${finalConfigMap.keys.join(', ')}");

      // Print each config key and its variation value only (match Kotlin)
      finalConfigMap.forEach((key, value) {
        if (value is Map<String, dynamic> && value.containsKey('variation')) {
          final variation = value['variation'];
          Logger.d("$key: $variation");
        } else {
          Logger.d("$key: $value");
        }
      });

      // Keep existing hero_text debug logging for backward compatibility
      try {
        final heroText = finalConfigMap['hero_text'] is Map<String, dynamic>
            ? (finalConfigMap['hero_text'] as Map<String, dynamic>)['variation']
            : null;
        if (heroText != null) {
          Logger.d("Hero text if present: $heroText");
        }
      } catch (e) {
        // Ignore error
      }

      return CFResult.success(finalConfigMap);
    } catch (e) {
      ErrorHandler.handleException(
        e,
        "Error parsing configuration response",
        source: _source,
        severity: ErrorSeverity.high,
      );

      return CFResult.error(
        "Error parsing configuration response: ${e.toString()}",
        exception: e,
        category: ErrorCategory.serialization,
      );
    }
  }

  /// Fetches metadata from a URL with improved error handling
  Future<CFResult<Map<String, String>>> fetchMetadata([String? url]) async {
    // If no URL is provided, construct the SDK settings URL (not user configs)
    final targetUrl = url ?? _buildSdkSettingsUrl();

    if (isOffline()) {
      Logger.d("Not fetching metadata because client is in offline mode");
      return CFResult.error("Client is in offline mode",
          category: ErrorCategory.network);
    }

    try {
      Logger.i(
          "API POLL: Fetch metadata strategy - First trying HEAD request: $targetUrl");

      // First try a lightweight HEAD request (match Kotlin)
      final headResult = await _httpClient.head(targetUrl);

      if (headResult.isSuccess) {
        final headers = headResult.getOrNull()?.headers.map ?? {};
        final metadata = {
          CFConstants.http.headerLastModified:
              headers['last-modified']?.first ?? '',
          CFConstants.http.headerEtag: headers['etag']?.first ?? '',
        };
        Logger.i("API POLL: HEAD request successful, using result: $metadata");
        return CFResult.success(metadata);
      }

      // If HEAD fails, fall back to the original GET method
      Logger.i(
          "API POLL: HEAD request failed, falling back to GET: $targetUrl");
      final getResult = await _httpClient.fetchMetadata(targetUrl,
          lastModified: _lastModified, etag: _lastEtag);

      if (getResult.isSuccess) {
        Logger.i("API POLL: Fallback GET successful: ${getResult.getOrNull()}");
      } else {
        Logger.w("API POLL: Both HEAD and GET failed for $targetUrl");
      }

      if (getResult.isSuccess) {
        // Store the headers for next request
        final headers = getResult.getOrNull() ?? {};
        _lastModified = headers[CFConstants.http.headerLastModified];
        _lastEtag = headers[CFConstants.http.headerEtag];
      }

      return getResult;
    } catch (e) {
      Logger.e(
          "API POLL: Exception during metadata fetch attempts: ${e.toString()}");
      ErrorHandler.handleException(
        e,
        "Error fetching metadata from $targetUrl",
        source: _source,
        severity: ErrorSeverity.high,
      );

      return CFResult.error(
        "Error fetching metadata: ${e.toString()}",
        exception: e,
        category: ErrorCategory.network,
      );
    }
  }

  /// Build SDK settings URL with dimension ID
  String _buildSdkSettingsUrl() {
    final String dimensionId = _config.dimensionId ?? "default";
    final sdkSettingsPath =
        CFConstants.api.sdkSettingsPathPattern.replaceFirst('%s', dimensionId);
    return CFConstants.api.sdkSettingsBaseUrl + sdkSettingsPath;
  }

  /// Returns the last successfully fetched configuration map
  CFResult<Map<String, dynamic>> getConfigs() {
    if (_lastConfigMap != null) {
      // Log the full last config map for easier debugging
      Logger.d('===== CONFIG FETCHER: LAST CONFIG MAP =====');
      _lastConfigMap!.forEach((key, value) {
        Logger.d('$key: $value');
      });
      Logger.d('=========================================');

      return CFResult.success(_lastConfigMap!);
    } else {
      return CFResult.error("No configuration has been fetched yet",
          category: ErrorCategory.validation);
    }
  }

  Future<CFResult<Map<String, dynamic>>> fetchSdkSettings() async {
    // Respect offline mode
    if (_offlineMode) {
      Logger.d('Not fetching SDK settings because client is in offline mode');
      return CFResult.error("Cannot fetch SDK settings in offline mode",
          category: ErrorCategory.network);
    }

    try {
      // Use the same URL building helper method for consistency
      final url = _buildSdkSettingsUrl();

      Logger.i("API POLL: Fetching full SDK settings with GET: $url");

      // Make request using fetchJson
      final result = await _httpClient.fetchJson(url);

      if (result.isSuccess) {
        Logger.i("API POLL: SDK settings parsed successfully");
      } else {
        Logger.w("API POLL: Failed to parse SDK settings response");
      }

      return result;
    } catch (e) {
      Logger.e(
          "API POLL: Exception during SDK settings fetch: ${e.toString()}");
      ErrorHandler.handleException(e, "Unexpected error fetching SDK settings",
          source: _source, severity: ErrorSeverity.high);

      return CFResult.error("Failed to fetch SDK settings",
          exception: e, category: ErrorCategory.internal);
    }
  }
}
